pkg/asset/ignition/bootstrap/baremetal/template.go

```diff
@@ -97,8 +111,109 @@ func externalURLs(apiVIPs []string) (externalURLv4 string, externalURLv6 string)
 	return
 }
 
+// TLSCertificate contains the private key and certificate contents in []byte format.
+type TLSCertificate struct {
+	privateKey  []byte
+	certificate []byte
+}
+
+func generateTLSCertificate(provisioningIP string) (TLSCertificate, error) {
+	var tlsExpirationDays = 1
+
+	caConfig, err := crypto.MakeSelfSignedCAConfig("metal3-ironic", tlsExpirationDays)
+	if err != nil {
+		return TLSCertificate{}, err
+	}
+
+	ca := crypto.CA{
+		Config:          caConfig,
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+	}
+
+	var host string
+	if provisioningIP == "" {
+		host = "localhost"
+	} else {
+		host = provisioningIP
+	}
+
+	config, err := ca.MakeServerCert(sets.New(host), tlsExpirationDays)
+
+	if err != nil {
+		return TLSCertificate{}, err
+	}
+
+	certBytes, keyBytes, err := config.GetPEMBytes()
+	if err != nil {
+		return TLSCertificate{}, err
+	}
+
+	return TLSCertificate{
+		privateKey:  keyBytes,
+		certificate: certBytes,
+	}, nil
+}
+
+func generateIronicVMediaTLSCertificate(provisioningIP string) error {
+	logrus.Debugf("------Before generating TLS certificate")
+	cert, err := generateTLSCertificate(provisioningIP)
+	if err != nil {
+		return err
+	}
+
+	logrus.Debugf("-------Writing TLS certificate contents for ironic to be used in case TLS for virtual media is enabled")
+	_, filename, _, ok := runtime.Caller(0) // /home/mdevscripts/installer/pkg/asset/manifests/openshift.go
+	if !ok {
+		errStr := "unable to get information about function invocations from the goroutine stack while creating TLS certificate"
+		logrus.Debugf(errStr)
+		return errors.New(errStr)
+	}
+
+	path := filepath.Dir(filepath.Dir(filepath.Dir(filepath.Dir(filename)))) + "/store/data/bootstrap/baremetal/files/etc/ironic"
+	logrus.Debugf("path: %s", path)
+	err = os.MkdirAll(path, 0777)
+	if err != nil {
+		logrus.Debugf("Error in creating directory: %s", err)
+	}
+
+	certfile, err := os.Create(path + "/ironic-virtualmedia-tls.crt")
+	if err != nil {
+		logrus.Debugf("-------Error in creating ironic-virtualmedia-tls.crt: %s", err)
+		return err
+	}
+	err = os.Truncate(path+"/ironic-virtualmedia-tls.crt", 0)
+	if err != nil {
+		logrus.Debugf("Error truncating ironic-virtualmedia-tls.crt file")
+	}
+	_, err = certfile.Write(cert.certificate)
+	if err != nil {
+		logrus.Debugf("-------Error in writing to ironic-virtualmedia-tls.crt: %s", err)
+		return err
+	}
+	certfile.Close()
```

Comment:
For the future: Close() should be used inside defer right after opening and checking the err.

Reply:
Will fix!

Reply:
Fixed